Protential memory leak.

sunheehnus opened the pull request:
queue.c/destroyQueue: release needjobs_responders and clients if needed

eg: GCQueue() will leak clients `if (q->clients && listLength(q->clients) == 0)`

antirez commented:
Thank you! Good catch, the list of responders was definitely leaked. For the clients, the matter is a bit more convoluted... Usually we should free queues only when there are no longer clients blocked into those queues, so this should be more an assert, basically! However there is another bug in `flushServerData` that calls `destroyQueue` without checking if there are clients attached.

So the right thing to do is likely:
1. Add the assert instead of the code to free the `clients` dictionary. However free it without asserting if it's just an empty dictionary.
2. Create a function that disconnects all the clients sending them an error `Queue deleted by the system administrator while the client was blocking on it` or alike, and use it in `flushServerData`.

Makes sense? Thanks again.

sunheehnus commented:
Hi @antirez .
Yes, make sense! I overlook the `flushServerData`. And sorry for the delay, feel illness for a week (:3J∠).

> Add the assert instead of the code to free the clients dictionary. However free it without asserting if it's just an empty dictionary.

``` c
--- a/src/queue.c
+++ b/src/queue.c
@@ -104,8 +104,10 @@ int destroyQueue(robj *name) {
     skiplistFree(q->sl);
     if (q->needjobs_responders)
         dictRelease(q->needjobs_responders);
-    if (q->clients)
+    if (q->clients) {
+        serverAssert(listLength(q->clients) == 0);
         listRelease(q->clients);
+    }
     zfree(q);
     return DISQUE_OK;
 }
```

> Create a function that disconnects all the clients sending them an error Queue deleted by the system administrator while the client was blocking on it or alike, and use it in flushServerData.

write a new function `unblockClientsBlockedOnQueue(queue *q)` in queue.c to disconnect and send error.

``` c
--- a/src/disque.c
+++ b/src/disque.c
@@ -527,6 +527,7 @@ void flushServerData(void) {
     dictSafeForeach(server.queues,de)
         queue *q = dictGetVal(de);
         serverAssert(queueLength(q) == 0);
+        unblockClientsBlockedOnQueue(q);
         destroyQueue(q->name);
     dictEndForeach
 }
--- a/src/queue.c
+++ b/src/queue.c
@@ -210,6 +212,25 @@ unsigned long queueNameLength(robj *qname) {
     return queueLength(lookupQueue(qname));
 }

+/* Unblock clients blocked on a queue */
+void unblockClientsBlockedOnQueue(queue *q) {
+    if (q->clients) {
+        listNode *ln;
+        listIter li;
+        listRewind(q->clients,&li);
+        while((ln = listNext(&li))) {
+            client *c = listNodeValue(ln);
+            unblockClient(c);
+            addReplyErrorFormat(c, "Queue %s deleted by the system"
+                    "administrator while the client was blocking on it",
+                    (char *)q->name->ptr);
+        }
+        listRelease(q->clients);
+        q->clients = NULL;
+    }
+}
+
+
 /* Remove a queue that was not accessed for enough time, has no clients
  * blocked, has no jobs inside. If the queue is removed DISQUE_OK is
  * returned, otherwise DISQUE_ERR is returned. */
--- a/src/queue.h
+++ b/src/queue.h
@@ -83,6 +83,7 @@ job *queueNameFetchJob(robj *qname, unsigned long *qlen);
 unsigned long queueLength(queue *q);
 unsigned long queueNameLength(robj *qname);
 void unblockClientBlockedForJobs(client *c);
+void unblockClientsBlockedOnQueue(queue *q);
 void handleClientsBlockedOnQueues(void);

 #define NEEDJOBS_CLIENTS_WAITING 0 /* Called because clients are waiting. */

```

Does this patch make sense? :)

sunheehnus commented:
Sorry, suddenly I realize that the `unblockClientsBlockedOnQueue` is not correct, later I will give a new version. :-)

sunheehnus commented:
Hi @antirez , this is the new `unblockClientsBlockedOnQueue()`
It iter all the clients blocked on the queue, unblock them and addReplyError as you recommend. 
Does this make sense? :-)

``` c
/* Unblock clients blocked on a queue */
void unblockClientsBlockedOnQueue(queue *q) {
    if (q && q->clients) {
        int numclients = listLength(q->clients);
        while(numclients--) {
            listNode *ln = listFirst(q->clients);
            client *c = ln->value;

            addReplyError(c, "Queue deleted by the system"
                    "administrator while blocking on it");
            unblockClient(c); /* This will remove it from q->clients. */
        }
    }
}
```

---

### ps: test on the new version

After client1 getjob from 123 234 and client2 getjob from 456, client3 flushall.

```
CLIENT1:
127.0.0.1:7711> GETJOB FROM 123 234
(error) ERR Queue deleted by the systemadministrator while blocking on it
(10.50s)
127.0.0.1:7711> PING
PONG

CLIENT2:
127.0.0.1:7711> GETJOB FROM 456
(error) ERR Queue deleted by the systemadministrator while blocking on it
(2.69s)
127.0.0.1:7711> PING
PONG

CLIENT3:
127.0.0.1:7711> DEBUG FLUSHALL
OK
127.0.0.1:7711> PING
PONG

```

antirez commented:
Thanks @sunheehnus, I'll merge ASAP, currently trying to figure out what's wrong with #73, reported after #72 (both crashes fixed for this issue hopefully). If you happen to have a few spare minutes I would love to get your point of view about how such a stack trace as #73 happens, since the dictionary should have different hashing methods.

sunheehnus committed:
destroyQueue: release needjobs_responders and clients if needed

Before destroy queues, we must guarantee that there is
no job queued and no client blocked on any of them.
Destroy all the jobs guarantees there is no job queued.
Unblock all the blocked clients guarantees no client is
blocking on any queue.
After the first two steps, it is safe to destroy all the
queues.

sunheehnus commented:
Hi @antirez , 
I find out that my former commits were not all correct and they mess with exported APIs.
I make the changes more clean and into one commit. :-)

antirez commented:
Thanks :-) I'll review once ready.

sunheehnus commented:
Hi @antirez , It is ready now. :-)

antirez commented:
Merged! Thanks.

antirez commented:
@sunheehnus changed just a couple of things in 9a57b51
